### PR TITLE
OpenMessaging API Specification

### PR DESCRIPTION
**Pull request by sijie**
The first API specification is based on "Producer-Topic-Subscription-Consumer" model. The API includes:

- Define the driver interfaces for creating clients to interact with OMS messaging systems
- Define the Message and MessageId interfaces
- Define the Producer interfaces
- Define the Consumer interfaces
- Define the Reader interfaces
- Define interfaces for configuration and stats

### CODE REVIEW

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+import io.openmessaging.api.producer.ProducerConfiguration;
+import io.openmessaging.api.reader.Reader;
+import io.openmessaging.api.reader.ReaderConfiguration;
+import io.openmessaging.exception.OMSException;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An OMS client that talks to OMS messaging services.
+ */
+public interface Client extends AutoCloseable {
```
Comment (zhouxinyu) [inline openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: `AutoCloseable` and `CompletableFuture` are specific APIs in Java platform, while we want to use very simple and common APIs  to describe our specification, to make the multilingual APIs looks consistent as much as possible~

Comment (zhouxinyu) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: And we already defined the Future and Promise interfaces, please refer to [Future](http://openmessaging.cloud/openmessaging-java/io/openmessaging/Future.html).

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    /**
+     * Asynchronously create a producer with given {@code ProducerConfiguration} for publishing on a specific topic
+     *
+     * @param topic
+     *            The name of the topic where to produce
+     * @param conf
+     *            The {@code ProducerConfiguration} object
+     * @return Future of the asynchronously created producer object
+     */
+    CompletableFuture<Producer> createProducerAsync(String topic, ProducerConfiguration conf);
```
Comment (zhouxinyu) [inline openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: We have defined the interface `MessagingAccessPoint` to create, update and list the producers/consumers, and the `MessagingAccessPoint` instance is generated from OMS driver URL.

And, IMO, there is no need to provide an async method to create clients, the clients will be created at preparation stage in most cases.

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+     * @param topic
+     *            The name of the topic where to read
+     * @param startMessageId
+     *            The message id where the reader will position itself. The first message returned will be the one after
+     *            the specified startMessageId
+     * @param conf
+     *            The {@code ReaderConfiguration} object
+     * @return The {@code Reader} object
+     */
+    Reader createReader(String topic, MessageId startMessageId, ReaderConfiguration conf) throws OMSException;
```
Comment (zhouxinyu) [inline openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: It seems that `Reader` is a concept from Pulsar?

File: openmessaging-api/src/main/java/io/openmessaging/api/ClientDriver.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+ *
+ * <p>The second part of the URL specifies the messaging access points. The brackets indicate that the extra access
+ * points are optional, but a correct OMS driver url needs at least one access point, which consists of hostname and
+ * port, like localhost:8081.
+ *
+ * <p>The last part of the URL specifies the namespace. A namespace wraps the OMS resources in an abstraction that
+ * makes it appear to the users within the namespace that they have their own isolated instance of the global OMS
+ * resources.
+ */
+public interface ClientDriver {
```
Comment (zhouxinyu) [inline openmessaging-api/src/main/java/io/openmessaging/api/ClientDriver.java]: Great idea to abstract an interface for the driver, we have an internal class [AccessPointURI](http://openmessaging.cloud/openmessaging-java/io/openmessaging/internal/AccessPointURI.html) to parse the URL.

File: openmessaging-api/src/main/java/io/openmessaging/api/MessageBuilder.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+public interface MessageBuilder {
+
+    /**
+     * Set the content of the message.
+     *
+     * @param data
+     *            array containing the payload
+     * @return message builder
+     */
+    default MessageBuilder setContent(byte[] data) {
```
Comment (zhouxinyu) [inline openmessaging-api/src/main/java/io/openmessaging/api/MessageBuilder.java]: IMO, the default method is not suitable in our spec~

File: openmessaging-api/src/main/java/io/openmessaging/api/consumer/Consumer.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+     */
+    void receive(MessageListener listener) throws OMSException;
+
+    /**
+     * Acknowledge the consumption of a single message
+     *
+     * @param message
+     *            The {@code Message} to be acknowledged
+     */
+    void ack(Message message) throws OMSException;
```
Comment (zhouxinyu) [inline openmessaging-api/src/main/java/io/openmessaging/api/consumer/Consumer.java]: We provide the ack mechanism in [MessageListener.Context](http://openmessaging.cloud/openmessaging-java/io/openmessaging/consumer/MessageListener.Context.html) rather than in `Consumer`, we could consider which one is better. 😄

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java (hunk 1 above)
Comment (merlimat) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: In my opinion the important part is that the same concepts (Client, Producer, Consumer, ...) are exposed in the API for all the supported languages. 

Every language has its own idiomatic way to do implement the concepts and we should follow that when adapting the the API concepts to that language. 

For example, in Java the native way is to use `CompletableFuture`. This is what Java programmers are used to. This is what other libraries are expecting or integrating with. 

When defining the API for Python, we then need to make sure the API is Python-like and friendly to Python programmer. For example, in Python the common way to do async APIs is to use callback functions instead of futures. We should follow that instead of pushing a Java like API to Python programmers.

Comment (merlimat) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: > And we already defined the Future and Promise interfaces, please refer to Future.

I most strongly recommend against inventing new `Future` and `Promise` classes. Java already provide a much better abstraction that is natively supported by many libraries and frameworks. 

By using the standard `CompletableFuture`, a Java developer can chain multiple asynchrous operations in a single complex future. If we were to do a non-standard Future, that would be impossible. 

```java
return source.getMessage()
                  .thenCombine(msg -> producer.sendAsync(msg))
                  .thenCombine(msgId -> database.put(msg.getKey(), "..something"));
```
The above is a fictitious example, but it serve to illustrate that, if we play in the idiomatic Java way, the API will fill both natural to Java programmer as well it will integrate with all the other libraries.

Comment (zhouxinyu) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: We can discuss the native and common API which is better for spec.

BTW, `CompletableFuture` is imported in JDK 1.8, which is not friendly for SDK, many legacy applications even use JDK 1.6~

Comment (sijie) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: @zhouxinyu : I agreed with Matteo. we should not reinvent `Future` in java api. we should use CompletableFuture to chain the futures in the asynchronous applications. If the goal is to make consistent, why not adding similar CompletableFuture in other languages, rather than giving up a good standard library in java. As what Matteo said, the most important thing is to make sure the model and api semantic are consistent between languages, not classes.

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java (hunk 2 above)
Comment (sijie) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: @zhouxinyu yes `Client` is what you called `MessagingAccessPoint`. but why not keep things simpler, call it `Client`. 

The async method are required when you are using a client to create multiple producers on-the-fly. you don't want to be blocked on creating a producer and consumer.

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java (hunk 3 above)
Comment (sijie) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: Reader is the `StreamingConsumer` which doesn't maintain any consumption state. it is a sequential `consumer` on `reading` messages. That's why it called `Reader`.

File: openmessaging-api/src/main/java/io/openmessaging/api/MessageBuilder.java (hunk 5 above)
Comment (sijie) [reply openmessaging-api/src/main/java/io/openmessaging/api/MessageBuilder.java]: any reason? it is a shortcut. but the implementation can still override it.

File: openmessaging-api/src/main/java/io/openmessaging/api/consumer/Consumer.java (hunk 6 above)
Comment (sijie) [reply openmessaging-api/src/main/java/io/openmessaging/api/consumer/Consumer.java]: check `ackCummulative` and `ack`. it makes more sense to put `ack` in consumer, when you want to `ack` and `ackCummulative`. If you put `ack` in MessageListener.Context, you can't do ack cumulatively. There should be an order implication on acknowledge.

**Comment by sijie** [comment]
@zhouxinyu : I saw the travis ci is built with multiple versions. I would suggest us getting rid of java6 and java7. We should take advantage of completablefuture and functional api in java8, rather than being ambitious on supporting all legacy language versions.

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java (hunk 1 above)
Comment (sijie) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: @zhouxinyu checkout my last comment. oracle already stopped public updates on Java 7 since 2015. There is no point to support java 7 or even earlier version. I would suggest supporting 8 and forward, and leveraging all the lambda functionalities there.

Comment (vongosling) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: @merlimat @sijie Client concept is not suitable for standard, It is a coarse-grained implementation model. We must remember what we are doing for these days, we hope to create standard, user-friendly API standard. we must jump out from our implementation. 


@zhouxinyu I also agree we'd not better reinvent another similar Naming concept, such as future and promise. But, if we want to strengthen the JDK implementation, we could create similar but not ambiguous concept.


 
AutoCloseable is created since JDK 7, and CompletableFuture is imported from JDK 8. IMO, It comes up another argument - backward compatible. I know oracle recommend JDK 9 in its JDK download page for long times. and JDK 7 is not even supported when we use 'brew cast 'in Macbook. But, many cloud organization, such as Ali-Cloud, Amazon(refer here, https://repo1.maven.org/maven2/com/amazonaws/aws-java-sdk-pom/1.11.228/aws-java-sdk-pom-1.11.228.pom) are all supporting JDK 6 in its client SDK. For this reason, I hope we could not import high version syntax feature into our standard.  But, if you want to support more awesome coding experience for developer, we could discuss how to reach this point :-)

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java (hunk 2 above)
Comment (vongosling) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: The same reason, I prefer fine-grained definiation :-)

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java (hunk 3 above)
Comment (vongosling) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: -1, the reason as I have listed in the previous discussion. our reinvent reader, future and promise are all not good practice.

File: openmessaging-api/src/main/java/io/openmessaging/api/ClientDriver.java (hunk 4 above)
Comment (vongosling) [reply openmessaging-api/src/main/java/io/openmessaging/api/ClientDriver.java]: I hate client concept, what is ClientDriver ?

**Comment by vongosling** [comment]
@sijie I suggest adding demo in the PR, some new concept seems confused, such as Client, ackCummulative, close async***,create***Async. How many steps if we want to create the client instance for produce and consume in your PR, Could we switch easily from the private implementation to opensource, even to cloud mode?

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java (hunk 1 above)
Comment (merlimat) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: @zhouxinyu Java7 was EOLd on late 2014. Large majority of companies have switched to Java8 long time ago, since there no more security updates on Java7 since then.

Comment (sijie) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: @vongosling 
```
Client concept is not suitable for standard, It is a coarse-grained implementation model. 
```
`Client` is the equivalent of [MessagingAccessPoint](https://github.com/openmessaging/openmessaging-java/blob/master/openmessaging-api/src/main/java/io/openmessaging/MessagingAccessPoint.java) here. I don't understand why it is not suitable for standard...

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java (hunk 2 above)
Comment (sijie) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: @vongosling @zhouxinyu please tell me what is the difference between MessagingAccessPoint and Client here. It is the instance that application used for creating producer, consumer.

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java (hunk 3 above)
Comment (sijie) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: I am fine with removing Reader for now, since we haven't reached to an agreement about streaming piece. I am fine with starting with simple producer and consumer interface. we can circle back.

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java (hunk 1 above)
Comment (merlimat) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: Calling `MessagingAccessPoint` is a bit confusing in the sense that it seems to be a "server" component rather than part of the client API and library.

File: openmessaging-api/src/main/java/io/openmessaging/api/ClientDriver.java (hunk 4 above)
Comment (sijie) [reply openmessaging-api/src/main/java/io/openmessaging/api/ClientDriver.java]: @vongosling the driver is the entry point for application to create a client implementation by resolving the service url.

**Comment by sijie** [comment]
@vongosling I would add examples and a implementation.

**Comment by merlimat** [comment]
> @sijie I suggest adding demo in the PR, some new concept seems confused, such as Client, ackCummulative, close async***,create***Async. How many steps if we want to create the client instance for produce and consume in your PR, Could we switch easily from the private implementation to opensource, even to cloud mode?

@vongosling Good idea about the examples to demonstrate the API itself. 

About the sync vs async versions of the methods. Having both versions doesn't mean to have "more steps" in the application code, it's just a way to give the choice to the application developers. 

Typically there are ~3 class of applications: 
 * Simple applications --> Always use sync API, very simple code and patterns
 * Applications that require publish throughput --> Use sync API to create producer, but then use `sendAsync()` to publish, for performance reasons
 * Applications that either are already inherently asynchronous or that need to deal with many topics --> Use async version of API

Bottom line, if you only have sync API for common operations like creating a producer, that would limit the scalability of application using the OpenMessaging API.

**Comment by sijie** [comment]
@zhouxinyu @vongosling I pushed the examples for the proposed API. I basically followed your existing applications in the samples. You can see the proposed API is much simpler than yours and it is unified. I am not sure why you don't like Client, it is exactly same concept as what you have as 'MessagingAccessPoint', but with a much simpler and well known name than 'MessagingAccessPoint'.

File: openmessaging-api-samples/src/main/java/io/openmessaging/examples/consumer/ConsumerExample.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            "oms:rocketmq://localhost:10911/us-east:namespace");
+
+        ConsumerConfiguration conf = new ConsumerConfiguration();
+        String topic = "path/to/topic";
+        String subscriberId = "poll_consumer";
+        final Consumer pollConsumer = client.subscribe(topic, subscriberId, conf);
+
+        // poll messages
+        Message msg;
+        while ((msg = pollConsumer.receive()) != null) {
```
Comment (merlimat) [inline openmessaging-api-samples/src/main/java/io/openmessaging/examples/consumer/ConsumerExample.java]: Ideally, if there's no timeout, `receive()` should block forever and never return `null`.

File: openmessaging-api-samples/src/main/java/io/openmessaging/examples/producer/ProducerExample.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        // write message sync
+        MessageBuilder builder = null; // provide the builder from
+        Message msg =
+            builder.setContent("Hellow OMS".getBytes(UTF_8)).build();
+
+        MessageId msgId = producer.send(msg);
+        System.out.println("Send sync message OK, message id is " + msgId);
+
+        // write message async
+        CompletableFuture<MessageId> result = producer.sendAsync(msg);
```
Comment (merlimat) [inline openmessaging-api-samples/src/main/java/io/openmessaging/examples/producer/ProducerExample.java]: As a general rule, I prefer to keep the success and the error handling separated. That also helps to avoid surprises when chaining multiple async operations since the whole chain will be failing without doing anything additional.

```java
producer.sendAsync(msg).thenAccept(messageId -> {
        System.out.println("Send async message OK, message id is " + messageId);
    }).exceptionally(ex -> {
        System.out.println("Send async message Failed, cause : " + ex);
    });

```

File: openmessaging-api-samples/src/main/java/io/openmessaging/examples/consumer/ConsumerExample.java (hunk 7 above)
Comment (sijie) [reply openmessaging-api-samples/src/main/java/io/openmessaging/examples/consumer/ConsumerExample.java]: yeah this is just dump example :)

File: openmessaging-api/src/main/java/io/openmessaging/api/Client.java (hunk 1 above)
Comment (merlimat) [reply openmessaging-api/src/main/java/io/openmessaging/api/Client.java]: @vongosling For what is worth, AWS Java SDK is based on Java8 : https://aws.amazon.com/sdk-for-java/

> We’re pleased to announce the Developer Preview of the AWS SDK for Java 2.0. The 2.0 
> version of the SDK is a major rewrite of the 1.11.x code base. It’s built on top of Java 8 and 
> adds several, frequently requested features, like support for non-blocking I/O and
> ....

**Comment by sijie** [comment]
ping @vongosling @zhouxinyu ?

**Comment by vongosling** [comment]
same answer to the another pr
